crates/ra_hir/src/adt.rs

```diff
@@ -71,14 +72,23 @@ fn get_def_id(
     loc.id(db)
 }
 
+sortedvec! {
```

Comment:
I would love to do it with macros and in fact this is what I tried first, but it didn't seem to work well with closures. I wrote a little about in the second part of this post: https://fullyfaithful.eu/sortedvec/

Reply:
In summary: we would need to store a closure for the key generation, but this make things extremely painful. The `F` in `SortedVec<T, F>` would propagate to everything touching it :-(

Reply:
I imagine the following API could work?

```
#[derive(Default)]
struct SortedVec<T> {
    items: Vec<T>,
}

impl<T> SortedVec<T> {
    fn insert_with(&mut self, value: T, cmp: impl Fn(&T, &T) -> Ordering) {
    }
    fn get_with(&self, key: impl Fn(&T) -> Ordering) -> Option<&T> {
        None
    }
}

impl<K: Ord, V> SortedVec<(K, V)> {
    fn insert(&mut self, key: K, value: V) {
        self.insert_with((key, value), |p1, p2| p1.0.cmp(&p2.0))
    }
    fn get(&self, key: &K) -> Option<&V> {
        self.get_with(|kv| key.cmp(&kv.0)).map(|kv| &kv.1)
    }
}

fn main() {
    let mut m: SortedVec<(i32, String)> = SortedVec::default();
    m.insert(92, "the answer".into());
    m.get(&92);
}
```

Reply:
This is a general "call site closure" pattern: you usually get more powerful APIs if you, instead of fixing the closure at the structure creation time, supply the closure at the call-site. See https://docs.rs/once_cell/0.1.7/once_cell/#general-purpose-lazy-evaluation for another example of this phenomenon.

Reply:
That doesn't protect an invariant though, does it? You could insert items with a different comparison fn than the one you're getting them with and you'd get bad results.

Reply:
Indeed it makes invariant violation possible, but it also unlocks more powerful APIs. For example, with `_with` API it is possible to write string interner like

```
struct Interner {
    data: Vec<String>,
    lookup: SortedVec<usize>,
}

impl Interner {
    fn intern(&mut self, text: &str) -> usize {
        if let Some(idx) = self.lookup.get_with(|idx| text.cmp(&self.data[idx])) {
            return idx
        }
        let idx = self.data.len();
        self.lookup.insert_with(idx, |i1, i2| self.data[i1].cmp(&self.data[i2]));
        idx
    }
    fn untern(&self, idx: usize) -> &str {
        &self.data[idx]
    }
}
```

The "add `Ord` at type/construction time" API makes such interner impossible, due to "struct borrows its own field" problem.

To enforce invariant in `K: Ord` key, one can define a newtype

```
struct SortedVecMap<K, V>(SortedVec<(K, V)>);
```